rfcs/0013-transition-caching.md

```diff
@@ -0,0 +1,40 @@
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+The goal here is to introduce 2 new caches to the system: the `Invalid_transition_cache`, and the `Unprocessed_transition_cache`.
+
+`Invalid_transition_cache` would be a bloom filter based cache which is updated whenever any form of validation fails on a transition. A bloom filter is a good fit since it remains a constant size while giving a probablistic check to see if an item is in the set or not. A bloom filter can produce false negatives for the existence of an item in the set, but this is considered ok since we are still performing validation on the transition after the `Invalid_transition_cache` is check. The check for this cache would be hoisted up to the toplevel `Transition_router` layer of the transition handling system, as this check is equally shared across both `Bootstrap_controller` and `Transition_frontier_controller`. The `Invalid_transition_cache` should also be persisted to disk. Depending on the cost of writing it to disk, it will either be written out on every update, or written out on an interval.
+
+`Unprocessed_transition_cache` is scoped more explicitly to the `Transition_frontier_controller`. The set stored in this cache represents the set of transitions which have been read from the network but have not yet been processed and added to the transition frontier. Since the lifetime of elements in the set are finite, the `Unprocessed_transition_cache` can be represented as a hash set. It will be the responsiblity of the transition validator to add items to this cache, and the responsibility of the processor to invalidate the cache once transitions are added to the transition frontier. Transitions which are determined to be invalid need to also be invalidated.
+
+In order to assist in ensuring that items in the cache are properly invalidated, I recommend the introduction of a `'a Cached.t` type which will track the state of the item in one or more caches. The `Cached` module would provide an interface for performing cache related actions and would track a boolean value representing whether or not the required actions have been performed. What's special about the `'a Cached.t` type is that it will have a custom finalization handler which will throw an exception if no cache actions have been performed by the time it is garbage collected. This exception is toggled by a debug flag. When the debug flag is off, the finalization handler will nearly log a message.
```

Comment:
I’m not sure if we’d want this to ever not cause a crash. The leak that this would cause would be hard to pinpoint to this system as the cause esp. via bug reports from our users — a stacktrace on the other hand is much nicer.

Reply:
> This exception is toggled by a debug flag. When the debug flag is off, the finalization handler will nearly log a message.

Doesn't that cover this? I only want it to crash so that we don't miss it because it's very important. A log can easily get missed and won't, for example, break CI if someone causes a bug in this.

Reply:
"nearly log a message" should be "merely log a message"?

Reply:
I would want a crash on users' systems as well, but maybe this is a larger discussion and a different RFC. I've worked on teams where we "avoided assertions so that the user doesn't have a broken app", but when your assertions fail, you DO have a broken app, the developers just don't get bug reports.